### ISSUE TITLE
Mutable Matrix Null Space Concern

### ISSUE THREAD

**Issue by davikrehalt** [Macaulay2/M2#2941]
If we calculate the nullspace of the matrix
`a=mutableMatrix{{0.0,1.0,2.0,3.0,0.0,0.0,0.0,0.0},{4.0,3.0,2.0,1.0,0.0,0.0,0.0,0.0},{0.0,0.0,0.0,0.0,1.0,2.0,3.0,4.0},{0.0,0.0,0.0,0.0,3.0,2.0,1.0,0.0}}`
using 
'nullSpace(a)'
it outputs
'i2 : nullSpace(a)

o2 = | -1 -2 0  0  |
     | 2  3  0  0  |
     | -1 0  0  0  |
     | 0  -1 0  0  |
     | 0  0  2  4  |
     | 0  0  .5 0  |
     | 0  0  -1 0  |
     | 0  0  0  -1 |'

which has the following property
i3 : a*nullSpace(a)

o3 = | 0 0 0 0  |
     | 0 0 0 0  |
     | 0 0 0 0  |
     | 0 0 6 12 |

o3 : MutableMatrix

which does not seem correct.

**Comment by d-torrance** [comment]
Very strange!  In the meantime, you can use `ker(Matrix)`, which seems to work much better:

```m2
i2 : mutableMatrix gens ker matrix a
-- warning: experimental computation over inexact field begun
--          results not reliable (one warning given per session)

o2 = | 1  2  0  0  |
     | -2 -3 0  0  |
     | 1  0  0  0  |
     | 0  1  0  0  |
     | 0  0  1  2  |
     | 0  0  -2 -3 |
     | 0  0  1  0  |
     | 0  0  0  1  |

o2 : MutableMatrix

i3 : a * oo

o3 = | 0 0 0           0            |
     | 0 0 0           0            |
     | 0 0 4.44089e-16 4.44089e-16  |
     | 0 0 0           -8.88178e-16 |

o3 : MutableMatrix
```

**Comment by davikrehalt** [comment]
Yes that works!  However, I think this should still be labeled as a bug!

**Comment by davikrehalt** [comment]
if you let me know where the bug it, I can try to take a look and try to fix it!

**Comment by d-torrance** [comment]
I think it's somewhere in the engine:

```m2
i1 : code (nullSpace, MutableMatrix)

o1 = -- code for method: nullSpace(MutableMatrix)
     /usr/share/Macaulay2/Core/mutablemat.m2:323:32-323:70: --source code:
     nullSpace(MutableMatrix) := (M) -> map(ring M, rawLinAlgNullSpace raw M)
```

Any time something starts with `raw`, it means that it's likely written in C++ in the `e` directory of the repository.  In this case:

https://github.com/Macaulay2/M2/blob/379c9634000dd87221dc353e1aaaca413f75c81c/M2/Macaulay2/e/interface/mutable-matrix.cpp#L824-L834

**Comment by davikrehalt** [comment]
Hmm, is 
`    void FFpackNullSpace(const DMat<CoeffRing> &mat, 
                         DMat<CoeffRing> &nullspace, 
                         bool right_side)
    {
        right_side = !right_side; // because FFPACK stores by rows, not by columns.

        typedef typename CoeffRing::ElementType ElementType;
        ElementType* N = newarray( ElementType, mat.n_rows() * mat.n_cols());    
        mat.copy_elems(mat.n_rows()*mat.n_cols(), N, 1, mat.get_array(), 1); 
    
        size_t nr = mat.n_rows();
        size_t nc = mat.n_cols();
    
        ElementType *nullspaceFFPACK = 0;
    
        size_t nullspace_dim;
        size_t nullspace_leading_dim;
    
        FFPACK::NullSpaceBasis(mat.ring().field(),
                               (right_side ? FFLAS::FflasRight : FFLAS::FflasLeft),
                               nc, nr, N, nr, nullspaceFFPACK, nullspace_leading_dim, nullspace_dim);
    
        std::cerr << "leading dim = " << nullspace_leading_dim << " and dim = " << nullspace_dim << std::endl;
        //NOTUSED? size_t nullspace_nrows = (right_side ? nc : nullspace_dim);
        if (right_side && nullspace_dim != nullspace_leading_dim)
          {
            std::cerr << "error: this should not happen!" << std::endl;
          }
        else if (!right_side && nullspace_leading_dim != nc)
          {
            std::cerr << "error: this should not happen either!" << std::endl;
          }
    
        if (right_side)
          nullspace.resize(nullspace_dim,nr);
        else
          nullspace.resize(nc,nullspace_dim);
    
        mat.copy_elems(nullspace.n_rows() * nullspace.n_cols(), nullspace.get_array(), 1, nullspaceFFPACK, 1); 
    
        delete [] nullspaceFFPACK;
    }
`
getting called by the code?

**Comment by mikestillman** [comment]
This is an interesting issue...!  I have always thought that the A = LU (or PLU) decomposition definition requires that U is in (non-reduced) row echelon order: the first non-zero elements in each row occur in increasing columns.  But this is *not* what lapack provides for us.  In fact, what lapack provides is not sufficient to compute a basis of the null space, it seems (not surprising, as numerically that appears to be less interesting?). A common way to compute a basis of the nullspace is to use a QR (if rank = column rank) or an SVD decomposition.  In the example in this issue, the rank is not the column rank.

Some possible solutions:

- Compute an SVD (sigma, U, Vt) of a, then take the appropriate submatrix of (the transpose of) Vt.  This is the most numerically stable, if I'm not mistaken.  But it can take longer time.  This is quite easy to implement at least.
- Compute an LU decomposition, where the U is as described above.  Then it is easy to read off the nullspace.  This is often fine, but it can be numerically bad, and the code currently in the c++ engine for this is not designed to be high performance.  It is more of a fallback for when there aren't better algorithms available.  This is also easy to program, since we have most of the features ready
- Disallow it.  Seems like a bad idea to me.

@LukeOeding @mahrud @d-torrance @antonleykin @ggsmith Suggestions for how I should fix this?(Note: this is for dense mutable matrices, we have no good sparse functionality available yet).

**Comment by mikestillman** [comment]
Here is what lapack gives when used in Macaulay2:
```
a=mutableMatrix{{0.0,1.0,2.0,3.0,0.0,0.0,0.0,0.0},{4.0,3.0,2.0,1.0,0.0,0.0,0.0,0.0},{0.0,0.0,0.0,0.0,1.0,2.0,3.0,4.0},{0.0,0.0,0.0,0.0,3.0,2.0,1.0,0.0}}
(P,L,U) = LUdecomposition a
```

gives

```
o1 = | 0 1 2 3 0 0 0 0 |
     | 4 3 2 1 0 0 0 0 |
     | 0 0 0 0 1 2 3 4 |
     | 0 0 0 0 3 2 1 0 |

o1 : MutableMatrix

i2 : (P,L,U) = LUdecomposition a

o2 = ({1, 0, 2, 3}, | 1 0 0 0 |, | 4 3 2 1 0 0 0 0 |)
                    | 0 1 0 0 |  | 0 1 2 3 0 0 0 0 |
                    | 0 0 1 0 |  | 0 0 0 0 1 2 3 4 |
                    | 0 0 0 1 |  | 0 0 0 0 3 2 1 0 |
```
